File: tests/python/test_project_operations.py
"""Exercise closed-file project operations through the native Python bindings."""

from __future__ import annotations

import os
import shutil
from pathlib import Path

import pytest


def _fixture() -> Path:
    configured = os.environ.get("LFS_PROJECT_OPERATIONS_FIXTURE")
    if configured:
        return Path(configured)
    for parent in Path(__file__).parents:
        candidate = parent / ".codex_tmp/am_concept/testhome/projects/bonsai.licht"
        if candidate.is_file():
            return candidate
    return Path(__file__).parents[2] / "missing-project.licht"


@pytest.fixture(scope="module")
def native_io():
    try:
        from lichtfeld import io
    except ImportError as error:
        pytest.skip(f"native lichtfeld.io is unavailable: {error}")
    return io


def test_closed_file_operations(native_io, tmp_path):
    source = _fixture()
    if not source.is_file():
        pytest.skip(f"operations fixture is unavailable: {source}")
    path = tmp_path / "copy.licht"
    shutil.copy2(source, path)

    canceled = native_io.verify_project_file(path, cancel=lambda: True)
    assert canceled.status is native_io.ProjectVerificationStatus.CANCELED
    verified = native_io.verify_project_file(path)
    assert verified.status is native_io.ProjectVerificationStatus.VERIFIED

    plan = native_io.plan_reduce_size(path)
    assert plan.physical_size > 0
    assert plan.input_commit_uuid
    reduced = native_io.reduce_size(
        path, {"drop_unbound_checkpoints": False, "drop_embedded_dataset": False}
    )
    assert reduced.card.physical_file_size > 0
    assert reduced.recovery_copy.is_file()

    restored = native_io.restore_save(path, 1, tmp_path / "restored.licht")
    assert restored.project_uuid != native_io.inspect_project_card(path).project_uuid

    png = b"\x89PNG\r\n\x1a\n"
    native_io.set_project_preview(path, png)
    native_io.set_project_license(path, "CC-BY-4.0", "Python test")
    titled = native_io.set_project_title(path, "Python operation copy")
    assert titled.title == "Python operation copy"
    assert native_io.inspect_project_details(path).license.identifier == "CC-BY-4.0"

    compacted = native_io.compact_project_file(path)
    assert "older save points" in compacted.diagnostic
    native_io.clear_project_license(path)
    assert native_io.inspect_project_details(path).license is None


def test_contents_removals_persist_until_compaction(native_io, tmp_path):
    import base64
    import json
    source = _fixture()
    if not source.is_file():
        pytest.skip(f'operations fixture is unavailable: {source}')
    path = tmp_path / 'contents.licht'
    shutil.copy2(source,path)
    native_io.set_project_title(path,'Before removal')
    native_io.set_project_license(path,'CC-BY-4.0','Credit: Studio')
    png=base64.b64decode('iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+aGNcAAAAASUVORK5CYII=')
    native_io.set_project_preview(path,png)
    details=native_io.inspect_project_details(path)
    current=details.card.generation
    with pytest.raises(Exception,match='current save'):
        native_io.reduce_size(path,dict(compact=False,drop_unbound_checkpoints=False,save_generation=current))
    oldest=details.save_history[0].generation
    size=path.stat().st_size
    reduced=native_io.reduce_size(path,dict(compact=False,drop_unbound_checkpoints=False,save_generation=oldest))
    assert reduced.recovery_copy.is_file() and reduced.bytes_reclaimed==0
    assert path.stat().st_size>=size
    pending=json.loads(native_io.inspect_project_details(path).manifest['contents_removals'])['rows']
    assert pending[0]['id']==f'save:{oldest}'
    with pytest.raises(Exception,match='removed'):
        native_io.restore_save(path,oldest,tmp_path/'removed-save.licht')
    native_io.reduce_size(path,dict(compact=False,drop_unbound_checkpoints=False,drop_thumbnail=True))
    assert not native_io.inspect_project_details(path).card.has_preview
    native_io.clear_project_license(path)
    details=native_io.inspect_project_details(path)
    assert details.license is None
    assert {r['kind'] for r in json.loads(details.manifest['contents_removals'])['rows']}=={'save','thumbnail','license'}
    assert native_io.verify_project_file(path).status is native_io.ProjectVerificationStatus.VERIFIED
    before=path.stat().st_size
    native_io.compact_project_file(path)
    assert path.stat().st_size<before
    assert 'contents_removals' not in native_io.inspect_project_details(path).manifest


def test_restore_replaces_same_path_with_new_identity(native_io,tmp_path):
    source=_fixture()
    if not source.is_file():pytest.skip(f'operations fixture is unavailable: {source}')
    path=tmp_path/'restore.licht'
    shutil.copy2(source,path)
    native_io.set_project_title(path,'Old title')
    old=native_io.inspect_project_card(path)
    native_io.set_project_title(path,'New title')
    restored=native_io.restore_save(path,old.generation,path)
    assert restored.project_uuid!=old.project_uuid
    assert restored.title=='Old title'
    assert native_io.verify_project_file(path).status is native_io.ProjectVerificationStatus.VERIFIED
    assert len(list(tmp_path.glob('*.licht')))==1
    assert len(list(tmp_path.glob('*.bak')))==1


def test_removing_bound_checkpoint_preserves_visible_model(native_io, tmp_path):
    import json

    configured = os.environ.get("LFS_PROJECT_CHECKPOINT_FIXTURE")
    source = Path(configured) if configured else _fixture().parent / "mrnf/stump/project.licht"
    if not source.is_file():
        pytest.skip(f"checkpoint fixture is unavailable: {source}")
    path = tmp_path / "checkpoint.licht"
    shutil.copy2(source, path)
    before = native_io.inspect_project_details(path)
    checkpoint = next(cp for cp in before.retained_checkpoints if cp.retained and cp.binds_scene_graph)

    result = native_io.reduce_size(path, {
        "compact": False,
        "drop_unbound_checkpoints": False,
        "checkpoint_uuid": str(checkpoint.instance_uuid),
    })
    after = native_io.inspect_project_details(path)
    assert result.checkpoints_removed == 1
    assert result.bytes_reclaimed == 0
    assert result.recovery_copy.is_file()
    assert after.scene_graph.training_node_id is None
    assert after.scene_graph.node_counts_by_type == before.scene_graph.node_counts_by_type
    assert not any(cp.retained and cp.instance_uuid == checkpoint.instance_uuid for cp in after.retained_checkpoints)
    assert any(str(part.fourcc) == "SPLT" and part.stored_bytes > 0 for part in after.chapters)
    removed = json.loads(after.manifest["contents_removals"])["rows"]
    assert any(row["id"] == f"checkpoint:{checkpoint.instance_uuid}" for row in removed)
    assert native_io.verify_project_file(path).status is native_io.ProjectVerificationStatus.VERIFIED


def test_rebinding_repeatedly_keeps_recovery_copies_out_of_catalog(native_io, tmp_path):
    configured = os.environ.get("LFS_PROJECT_CHECKPOINT_FIXTURE")
    source = Path(configured) if configured else _fixture().parent / "mrnf/stump/project.licht"
    if not source.is_file():
        pytest.skip(f"checkpoint fixture is unavailable: {source}")
    path = tmp_path / "resume.licht"
    shutil.copy2(source, path)
    details = native_io.inspect_project_details(path)
    checkpoint = next(cp for cp in details.retained_checkpoints if cp.retained and cp.binds_scene_graph)
    first = native_io.rebind_checkpoint(path, str(checkpoint.instance_uuid))
    second = native_io.rebind_checkpoint(path, str(checkpoint.instance_uuid))
    assert first.commit_uuid != second.commit_uuid
    assert first.project_uuid == second.project_uuid == details.card.project_uuid
    assert len(list(tmp_path.glob("*.licht"))) == 1
    assert len(list(tmp_path.glob("*.bak"))) == 2
    assert native_io.verify_project_file(path).status is native_io.ProjectVerificationStatus.VERIFIED


def test_chosen_thumbnail_survives_other_contents_edits(native_io, tmp_path):
    import struct
    import zlib

    def png(rgb):
        def chunk(kind, payload):
            return struct.pack(">I", len(payload)) + kind + payload + struct.pack(">I", zlib.crc32(kind + payload))

        return (b"\x89PNG\r\n\x1a\n"
                + chunk(b"IHDR", struct.pack(">IIBBBBB", 2, 1, 8, 2, 0, 0, 0))
                + chunk(b"IDAT", zlib.compress(b"\x00" + bytes(rgb) * 2))
                + chunk(b"IEND", b""))

    source = _fixture()
    if not source.is_file():
        pytest.skip(f"operations fixture is unavailable: {source}")
    path = tmp_path / "preview.licht"
    shutil.copy2(source, path)
    dataset = tmp_path / "dataset"
    for folder in ("images", "images_4"):
        images = dataset / folder
        images.mkdir(parents=True)
        (images / "first.png").write_bytes(png((255, 0, 0)))
    native_io.set_dataset_reference(path, dataset, True)
    native_io.preview_from_first_dataset_image(path)
    chosen = png((0, 0, 255))
    assert native_io.read_preview(path) != chosen

    native_io.set_project_preview(path, chosen)
    assert native_io.read_preview(path) == chosen
    native_io.set_project_license(path, "CC-BY-4.0", "Credit: Studio")
    assert native_io.read_preview(path) == chosen
    native_io.reduce_size(path, {
        "compact": False,
        "drop_unbound_checkpoints": False,
        "drop_thumbnail": True,
    })
    assert not native_io.inspect_project_card(path).has_preview
    native_io.clear_project_license(path)
    assert not native_io.inspect_project_card(path).has_preview
    assert native_io.verify_project_file(path).status is native_io.ProjectVerificationStatus.VERIFIED

File: tests/test_project_operations.cpp
/* SPDX-FileCopyrightText: 2026 LichtFeld Studio Authors
 * SPDX-License-Identifier: GPL-3.0-or-later */

#include "core/checkpoint_format.hpp"
#include "io/project_document.hpp"
#include "io/project_operations.hpp"
#include "licht_test_support.hpp"

#include <algorithm>
#include <cstring>
#include <filesystem>
#include <gtest/gtest.h>

namespace {

    namespace fs = std::filesystem;
    using namespace lfs::io::project;
    using namespace lfs::test::licht;

    std::vector<std::byte> checkpoint_payload(const std::int32_t iteration) {
        lfs::core::CheckpointHeader header{};
        header.iteration = iteration;
        header.num_gaussians = 4;
        header.sh_degree = 2;
        std::vector<std::byte> result(sizeof(header));
        std::memcpy(result.data(), &header, sizeof(header));
        return result;
    }

    lfs::Result<ProjectDocumentSaveReport> save_document(
        ProjectDocument& document, const fs::path& path,
        const lfs::core::Uuid& file_uuid = fixed_uuid(2000)) {
        ProjectDocumentSaveOptions options;
        options.file_uuid = file_uuid;
        options.index_compression = IndexCompression::StoredForDeterministicTests;
        options.disk_reserve_bytes = 0;
        return document.save(path, options);
    }

    fs::path make_document(const fs::path& path) {
        auto document = require_result(ProjectDocument::create(fixed_uuid(1000), 1'700'000'000'000'000'000));
        static_cast<void>(require_result(save_document(document, path)));
        return path;
    }

    TEST(ProjectOperations, RestoreOlderSaveRekeysAndRefusesCollision) {
        TemporaryDirectory temporary;
        const auto source = make_document(temporary.path / "source.licht");
        auto document = require_result(ProjectDocument::open(source));
        static_cast<void>(document.edit_project().dom().set("old_marker", "old"));
        static_cast<void>(require_result(document.save(source, ProjectDocumentSaveOptions{
                                                                   .index_compression = IndexCompression::StoredForDeterministicTests,
                                                                   .disk_reserve_bytes = 0,
                                                               })));
        const auto destination = temporary.path / "restored.licht";
        const auto card = require_result(restore_save(source, 1, destination));
        const auto source_card = require_result(inspect_project_card(source));
        EXPECT_NE(card.project_uuid, source_card.project_uuid);
        EXPECT_NE(card.file_uuid, source_card.file_uuid);
        EXPECT_EQ(card.generation, 1u);
        EXPECT_FALSE(require_result(inspect_project_details(destination)).card.title.has_value());
        const auto collision = restore_save(source, 1, destination);
        ASSERT_FALSE(collision);
        EXPECT_EQ(collision.error().code(), lfs::ErrorCode::AlreadyExists);
    }

    TEST(ProjectOperations, RebindCheckpointKeepsRecoveryCopy) {
        TemporaryDirectory temporary;
        const auto path = temporary.path / "resume.licht";
        auto document = require_result(ProjectDocument::create(fixed_uuid(1100), 1'700'000'000'000'000'000));
        const auto first = fixed_uuid(1101);
        const auto second = fixed_uuid(1102);
        require_status(document.set_checkpoint(
            first, require_result(LazyChunkValue::from_owned(checkpoint_payload(10), first))));
        require_status(document.set_checkpoint(
            second, require_result(LazyChunkValue::from_owned(checkpoint_payload(20), second))));
        SceneNodeRecord node;
        node.uuid = fixed_uuid(1110);
        node.type = "splat";
        node.name = "Training model";
        node.payload = PayloadBinding{
            .fourcc = "CKPT",
            .instance_uuid = second,
            .source_kind = "checkpoint"};
        require_status(document.edit_scene_graph().upsert_node(node));
        require_status(document.edit_scene_graph().set_training_model_uuid(node.uuid));
        static_cast<void>(require_result(save_document(document, path, fixed_uuid(1111))));

        const auto card = require_result(rebind_checkpoint(path, first));
        EXPECT_EQ(card.open_state, OpenState::Open);
        EXPECT_TRUE(fs::is_regular_file(path.string() + ".before-rebind.licht"));
        const auto details = require_result(inspect_project_details(path));
        ASSERT_TRUE(details.scene_graph.training_node_id.has_value());
        ASSERT_TRUE(details.retained_checkpoints.size() >= 2);
        const auto bound = std::ranges::find_if(
            details.retained_checkpoints,
            [](const auto& checkpoint) { return checkpoint.binds_scene_graph; });
        ASSERT_NE(bound, details.retained_checkpoints.end());
        EXPECT_EQ(bound->instance_uuid, first);
    }

    TEST(ProjectOperations, CompactPreservesRetainedCheckpointsAndWarns) {
        TemporaryDirectory temporary;
        const auto path = temporary.path / "compact.licht";
        auto document = require_result(ProjectDocument::create(fixed_uuid(1200), 1'700'000'000'000'000'000));
        const auto first = fixed_uuid(1201);
        const auto second = fixed_uuid(1202);
        require_status(document.set_checkpoint(
            first, require_result(LazyChunkValue::from_owned(checkpoint_payload(10), first))));
        require_status(document.set_checkpoint(
            second, require_result(LazyChunkValue::from_owned(checkpoint_payload(20), second))));
        static_cast<void>(require_result(save_document(document, path, fixed_uuid(1210))));
        const auto before = require_result(inspect_project_details(path));
        const auto compacted = require_result(compact_project_file(path));
        EXPECT_NE(compacted.diagnostic.find("older save points"), std::string::npos);
        const auto after = require_result(inspect_project_details(path));
        ASSERT_EQ(after.retained_checkpoints.size(), before.retained_checkpoints.size());
        EXPECT_EQ(after.card.commit_kind, CommitKind::Compaction);
        EXPECT_EQ(after.save_history.size(), 1u);
    }

    TEST(ProjectOperations, VerifyCanBeCanceledAndPreviewLicenseTitleAreExplicitSaves) {
        TemporaryDirectory temporary;
        const auto path = make_document(temporary.path / "metadata.licht");
        auto canceled = require_result(verify_project_file(
            path, {}, [] { return true; }));
        EXPECT_EQ(canceled.status, ProjectVerificationStatus::Canceled);
        auto verified = require_result(verify_project_file(path));
        EXPECT_EQ(verified.status, ProjectVerificationStatus::Verified);
        const std::vector<std::byte> preview{std::byte{'p'}, std::byte{'n'}, std::byte{'g'}};
        static_cast<void>(require_result(set_project_preview(path, preview)));
        static_cast<void>(require_result(set_project_license(path, "CC-BY-4.0", "Notice")));
        auto titled = require_result(set_project_title(path, "A useful title"));
        ASSERT_TRUE(titled.title.has_value());
        EXPECT_EQ(*titled.title, "A useful title");
        auto details = require_result(inspect_project_details(path));
        ASSERT_TRUE(details.license.has_value());
        EXPECT_EQ(details.license->identifier, "CC-BY-4.0");
        static_cast<void>(require_result(clear_project_license(path)));
        details = require_result(inspect_project_details(path));
        EXPECT_FALSE(details.license.has_value());
    }

    TEST(ProjectOperations, MutationsUseTheClosedFileWriterLockMessage) {
        TemporaryDirectory temporary;
        const auto path = make_document(temporary.path / "locked.licht");
        auto lease = require_result(WriterLockLease::acquire(path));
        const auto result = set_project_title(path, "blocked");
        ASSERT_FALSE(result);
        EXPECT_EQ(result.error().code(), lfs::ErrorCode::Unavailable);
        EXPECT_EQ(result.error().user_message(),
                  "The project is open for writing in another LichtFeld Studio");
    }

} // namespace
